### ISSUE TITLE
Emscripten: machine type must be wasm32 or wasm64

### ISSUE THREAD

**Issue by cuinjune** [openframeworks/openFrameworks#6615]
If I build for emscripten (using the latest version), I get the below errors.

It seems like we need to rebuild these libraries (according to https://github.com/emscripten-core/emscripten/issues/10342)

I wonder if there's any Makefile or script to rebuild these libraries.

```
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(BitmapAccess.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(GetType.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(MemoryIO.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Plugin.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Conversion.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Conversion24.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Conversion32.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Conversion8.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(ConversionType.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Flip.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/FreeImage/lib/emscripten/libfreeimage.a(Rescale.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/boost/lib/emscripten/libboost_system.a(error_code.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/boost/lib/emscripten/libboost_filesystem.a(operations.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/boost/lib/emscripten/libboost_filesystem.a(path.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/freetype/lib/emscripten/libfreetype.bc: machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/pugixml/lib/emscripten/libpugixml.bc: machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/tess2/lib/emscripten/libtess2.a(tess.c.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/uriparser/lib/emscripten/liburiparser.a(UriFile.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/uriparser/lib/emscripten/liburiparser.a(UriNormalize.o): machine type must be wasm32 or wasm64
wasm-ld: error: /Users/cuinjune/Downloads/openFrameworks-patch-release/libs/uriparser/lib/emscripten/liburiparser.a(UriParse.o): machine type must be wasm32 or wasm64
```

**Comment by arturoc** [comment]
This libraries are rebuild from the apothecary repository everytime there's a new commit. You can send a new PR there to test that building would work correctly and we can merge it once all the tests pass

**Comment by cuinjune** [comment]
I'm not sure if this is what you meant but I made a new PR: https://github.com/openframeworks/apothecary/pull/158

**Comment by arturoc** [comment]
Please test this with tomorrows nightly builds and close the issue if it's already solved by then or report back if it isn't.

Thanks!

**Comment by cuinjune** [comment]
@arturoc Thanks! I just tried again with the latest nightly build on macOS but I'm getting different errors.

Here's the last part of the log:

```
emcc: warning: generating an executable with an object extension (.bc).  If you meant to build an object file please use `-c, `-r`, or `-shared` [-Wemcc]
error: undefined symbol: _ZN24ofxEmscriptenSoundStreamC1Ev (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: __ZN24ofxEmscriptenSoundStreamC1Ev may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN24ofxEmscriptenSoundStreamD1Ev (referenced by top-level compiled C/C++ code)
warning: __ZN24ofxEmscriptenSoundStreamD1Ev may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN5boost6system15system_categoryEv (referenced by top-level compiled C/C++ code)
warning: __ZN5boost6system15system_categoryEv may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN5boost6system16generic_categoryEv (referenced by top-level compiled C/C++ code)
warning: __ZN5boost6system16generic_categoryEv may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: tessDeleteTess (referenced by top-level compiled C/C++ code)
warning: _tessDeleteTess may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: tessNewTess (referenced by top-level compiled C/C++ code)
warning: _tessNewTess may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/Users/cuinjune/emsdk/node/12.18.1_64bit/bin/node /Users/cuinjune/emsdk/upstream/emscripten/src/compiler.js /var/folders/zh/4x4533vs5vs1cqpyk5ch945w0000gn/T/tmp1hldoqzp.txt' failed (1)
make[2]: *** [/Users/cuinjune/Downloads/of_v20200922_osx_release/libs/openFrameworksCompiled/lib/emscripten/libopenFrameworks.bc] Error 1
make[1]: *** [Release] Error 2
make: *** [Release] Error 2
```
I tried adding `-s ERROR_ON_UNDEFINED_SYMBOLS=0`  to `PLATFORM_OPTIMIZATION_CFLAGS` and `PLATFORM_OPTIMIZATION_LDFLAGS` in `config.emscripten.default.mk` but I still get the same errors.

Any suggestions to fix this error?

**Comment by SimplyOnMyWay** [comment]
hi I'm happy to test / help with this issue if there's anything I can do.  I am unable to find a way to get emscripten to work with the linux release of OF (of_v20200924_linux64gcc6_release).  My OS is Ubuntu 20.04.  Have tried as many versions of emscripten as possible but seem to be going down a rabbit hole.  First time offering help with an issue. Let me know if my help isn't needed or I'm conflating my issue with a separate one here.  I'm learning here - keep me right, thanks!

**Comment by cuinjune** [comment]
@SimplyOnMyWay Your help would of course be great! 
Here's the script I run to install emscripten:
```
cd ~/emsdk && git pull && ./emsdk install latest && ./emsdk activate latest && source ./emsdk_env.sh
./emsdk update-tags && ./emsdk install tot && ./emsdk activate tot
```
I tried it with and without the second line but they both generate the [errors](https://github.com/openframeworks/openFrameworks/issues/6615#issuecomment-696577032)

**Comment by SimplyOnMyWay** [comment]
hi I'm happy to test / help with this issue if there's anything I can do.  I am unable to find a way to get emscripten to work with the linux release of OF (of_v20200924_linux64gcc6_release).  My OS is Ubuntu 20.04.  Have tried as many versions of emscripten as possible but seem to be going down a rabbit hole.  First time offering help with an issue. Let me know if my help isn't needed or I'm conflating my issue with a separate one here.  I'm learning here - keep me right, thanks!


> @SimplyOnMyWay Your help would of course be great!
> Here are the commands I run to install emscripten:
> 
> ```
> cd ~/emsdk && git pull && ./emsdk install latest && ./emsdk activate latest && source ./emsdk_env.sh
> ./emsdk update-tags && ./emsdk install tot && ./emsdk activate tot
> ```
> 
> I tried it with and without the second command line but they both generate the same [errors](https://github.com/openframeworks/openFrameworks/issues/6615#issuecomment-696577032).
> 
> Have you tried the same method to install emscripten?
> And then you can try running `emmake make` inside `OF/examples/3d/3DPrimitivesExample` directory.

hi @cuinjune!  I tried your commands above, and got what seems like line for line identical errors to what you posted above (3 days ago).  

Previously I had followed the "https://forum.openframeworks.cc/t/latest-emscripten-doesnt-work-on-examples/30463/15" thread on the forum.  This thread led to some positive results for some, but not for others - the positive lessons seem to have been summarised by @ofTheo here: "https://github.com/openframeworks/ofSite/issues/716".   From the latter, I followed these commands to the letter:
./emsdk install emscripten-1.38.9
./emsdk install sdk-1.38.9-64bit
./emsdk install binaryen-tag-1.38.9-64bit
./emsdk install clang-e1.38.9-64bit

./emsdk activate emscripten-1.38.9
./emsdk activate sdk-1.38.9-64bit
./emsdk activate binaryen-tag-1.38.9-64bit
./emsdk activate clang-e1.38.9-64bit

source emsdk_env.sh

This did actually compile for me.  At least I think it did, from memory.  In addition to these 9 lines, I also downloaded and ran the script download_libs.sh, which was missing from my OF/scripts/emscripten folder.  I have no idea if this is important - perhaps its a legacy script superceded elsewhere.  I got it and the OF/scripts/dev folder on which it is dependent from @moebiussurfing's Dec 19 comment on the first forum thread, ie
https://forum.openframeworks.cc/uploads/short-url/s4yYGeHHYjCAQgV2OYXhcWlHPei.zip
https://forum.openframeworks.cc/uploads/short-url/4niZt85kgOUG3IVrFgcBArrJ52C.zip

It seems this may not be necessary - I don't know what I'm doing!!!

**Comment by SimplyOnMyWay** [comment]
> hi I'm happy to test / help with this issue if there's anything I can do. I am unable to find a way to get emscripten to work with the linux release of OF (of_v20200924_linux64gcc6_release). My OS is Ubuntu 20.04. Have tried as many versions of emscripten as possible but seem to be going down a rabbit hole. First time offering help with an issue. Let me know if my help isn't needed or I'm conflating my issue with a separate one here. I'm learning here - keep me right, thanks!
> 
> > @SimplyOnMyWay Your help would of course be great!
> > Here are the commands I run to install emscripten:
> > ```
> > cd ~/emsdk && git pull && ./emsdk install latest && ./emsdk activate latest && source ./emsdk_env.sh
> > ./emsdk update-tags && ./emsdk install tot && ./emsdk activate tot
> > ```
> > 
> > 
> > I tried it with and without the second command line but they both generate the same [errors](https://github.com/openframeworks/openFrameworks/issues/6615#issuecomment-696577032).
> > Have you tried the same method to install emscripten?
> > And then you can try running `emmake make` inside `OF/examples/3d/3DPrimitivesExample` directory.
> 
> hi @cuinjune! I tried your commands above, and got what seems like line for line identical errors to what you posted above (3 days ago).
> 
> Previously I had followed the "https://forum.openframeworks.cc/t/latest-emscripten-doesnt-work-on-examples/30463/15" thread on the forum. This thread led to some positive results for some, but not for others - the positive lessons seem to have been summarised by @ofTheo here: "[openframeworks/ofSite#716](https://github.com/openframeworks/ofSite/issues/716)". From the latter, I followed these commands to the letter:
> ./emsdk install emscripten-1.38.9
> ./emsdk install sdk-1.38.9-64bit
> ./emsdk install binaryen-tag-1.38.9-64bit
> ./emsdk install clang-e1.38.9-64bit
> 
> ./emsdk activate emscripten-1.38.9
> ./emsdk activate sdk-1.38.9-64bit
> ./emsdk activate binaryen-tag-1.38.9-64bit
> ./emsdk activate clang-e1.38.9-64bit
> 
> source emsdk_env.sh
> 
> This did actually compile for me. At least I think it did, from memory. In addition to these 9 lines, I also downloaded and ran the script download_libs.sh, which was missing from my OF/scripts/emscripten folder. I have no idea if this is important - perhaps its a legacy script superceded elsewhere. I got it and the OF/scripts/dev folder on which it is dependent from @moebiussurfing's Dec 19 comment on the first forum thread, ie
> https://forum.openframeworks.cc/uploads/short-url/s4yYGeHHYjCAQgV2OYXhcWlHPei.zip
> https://forum.openframeworks.cc/uploads/short-url/4niZt85kgOUG3IVrFgcBArrJ52C.zip
> 
> It seems this may not be necessary - I don't know what I'm doing!!!

Unfortunately while it did compile, the javascript and wasm code isn't working:
https://michealocathain.com/av/emscripten_tests/3DPrimitives/3DPrimitivesExample.html

I'm still trying to read through the solutions to this which seem to have been figured out by @stephanschultz and other in Dec 19 at https://forum.openframeworks.cc/t/latest-emscripten-doesnt-work-on-examples/30463/37.  Sorry not to be doing anything more than redirecting to forum posts you may well have read and absorbed already!

**Comment by SimplyOnMyWay** [comment]
Latest test this morning (26.09.20) on Ubuntu 20.04, with fresh download of nightly build of OF of_v20200925_linux64gcc6_release. 

Preliminary steps to setup OF:  ran following commands:
of_v20200925_linux64gcc6_release/scripts/linux/ubuntu/install_dependencies.sh
of_v20200925_linux64gcc6_release/scripts/linux/ubuntu/install_codecs.sh
of_v20200925_linux64gcc6_release/scripts/linux/compileOF.sh

Preliminary steps to setup emscripten:
git clone https://github.com/emscripten-core/emsdk.git
cd emsdk
git pull

These next steps seems to be critical:
My basis for what appears to be the latest "best practice" is this post (albeit from Dec 2019):
https://github.com/openframeworks/ofSite/issues/716#issuecomment-560988470
This post recommends:
/emsdk install sdk-incoming-64bit binaryen-master-64bit
 
The /emsdk repository I cloned does not seem to include sdk-incoming-64bit.
It does include binaryen-master-64bit.
So, I have substituted sdk-upstream-master-64bit
***This last substitution could be introducing problems***

So, my commands were:
./emsdk install sdk-upstream-master-64bit binaryen-master-64bit
./emsdk activate sdk-upstream-master-64bit binaryen-master-64bit

NOTE: installing binaryen-master-64bit this way builds from source, and required an overnight build!
There could well be a much quicker way to do this via binaries.  
Have not checked yet, but expect it may be possible depending on what is returned by ./emsdk list --old. 


Following install and activation of ./emsdk toolset as above, I navigated to 
~/of_v20200925_linux64gcc6_release/examples/3d/3DPrimitivesExample
and ran command
emmake make -j2

The last messages of the compile appear identical to what you posted above @cuinjune :1


cache:INFO: generating system asset: generated_struct_info.json... (this will be cached in "/home/ubuntu/emsdk/emscripten/master/cache/wasm/generated_struct_info.json" for subsequent builds)
cache:INFO:  - ok
error: undefined symbol: _ZN24ofxEmscriptenSoundStreamC1Ev (referenced by top-level compiled C/C++ code)
warning: Link with `-s LLD_REPORT_UNDEFINED` to get more information on undefined symbols
warning: To disable errors for undefined symbols use `-s ERROR_ON_UNDEFINED_SYMBOLS=0`
warning: __ZN24ofxEmscriptenSoundStreamC1Ev may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN24ofxEmscriptenSoundStreamD1Ev (referenced by top-level compiled C/C++ code)
warning: __ZN24ofxEmscriptenSoundStreamD1Ev may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN5boost6system15system_categoryEv (referenced by top-level compiled C/C++ code)
warning: __ZN5boost6system15system_categoryEv may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: _ZN5boost6system16generic_categoryEv (referenced by top-level compiled C/C++ code)
warning: __ZN5boost6system16generic_categoryEv may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: tessDeleteTess (referenced by top-level compiled C/C++ code)
warning: _tessDeleteTess may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
error: undefined symbol: tessNewTess (referenced by top-level compiled C/C++ code)
warning: _tessNewTess may need to be added to EXPORTED_FUNCTIONS if it arrives from a system library
Error: Aborting compilation due to previous errors
emcc: error: '/home/ubuntu/emsdk/node/12.18.1_64bit/bin/node /home/ubuntu/emsdk/emscripten/master/src/compiler.js /tmp/tmp8vpdxe2z.txt' failed (1)
make[2]: *** [makefileCommon/compile.core.mk:264: /home/ubuntu/of_v20200925_linux64gcc6_release/libs/openFrameworksCompiled/lib/emscripten/libopenFrameworks.bc] Error 1
make[1]: *** [makefileCommon/compile.core.mk:204: Release] Error 2
make[1]: Leaving directory '/home/ubuntu/of_v20200925_linux64gcc6_release/libs/openFrameworksCompiled/project'
make: *** [/home/ubuntu/of_v20200925_linux64gcc6_release/libs/openFrameworksCompiled/project/makefileCommon/compile.project.mk:125: Release] Error 2

**Comment by ofTheo** [comment]
Apologies for emscripten been broken. 
Could you see if it works for you now with the nightly builds ( bottom of https://openframeworks.cc/download/ ) and using the new [setup instructions here](https://openframeworks.cc/setup/emscripten/ ) 

I believe this is fixed now. 
If not please comment and we can reopen it. Thanks!

**Comment by SimplyOnMyWay** [comment]
> Apologies for emscripten being broken.
> Could you see if it works for you now with the nightly builds ( bottom of https://openframeworks.cc/download/ ) and using the new [setup instructions here](https://openframeworks.cc/setup/emscripten/)
> 
> I believe this is fixed now.
> If not please comment and we can reopen it. Thanks!

Hi @ofTheo !  Thanks for following this up.  I have just rerun everything with last nights build and a reinstall of emscritpen following the latest guide.  Unfortunately I am getting an "exception thrown, see Javascript console" error, which from memory is the same as before.  

Let me know if I can help troubleshoot!

**Comment by ofTheo** [comment]
Hmm I just tried again with the latest builds and it seems to work for me ( on macOS 10.14 ). 
I just tried with a clean install.  

1. Deleted my emsdk directory.
2. Removed `source "/Users/theo/emsdk/emsdk_env.sh"` from my .bash_profile  
3. Close my current terminal and reopen.
4. `git clone https://github.com/emscripten-core/emsdk`
5. `cd emsdk`
6. `./emsdk install sdk-1.40.0-64bit`
7. `./emsdk activate sdk-1.40.0-64bit` 
8. `echo 'source "/Users/theo/emsdk/emsdk_env.sh"' >> $HOME/.bash_profile`
9. close my terminal and reopen 
10. Download nightly build ( of_v20210130_osx_release ) 
11. `cd of_v20210130_osx_release/examples/3d/3DPrimitivesExample/`
12. `emmake make`
13. `emrun --browser chrome bin/3DPrimitivesExample.html`

If the above doesn't work for you could you share the error messages?

**Comment by SimplyOnMyWay** [comment]
Hi @ofTheo 

Thanks I went through each step as you laid them out, but no joy unfortunately.
Screengrab of console.log here: https://michealocathain.com/7rl/Screenshot%20from%202021-02-01%2013-01-51.png

**Comment by ofTheo** [comment]
Thanks @SimplyOnMyWay 
Weird as it looks like it builds without issue. 

Could you try for the last line:

`emrun --no_browser --port 8080 bin/3DPrimitivesExample.html`

Then open a webgl compatible browser and go to this location `localhost:8080` 

I wonder if it could be a browser extension or other extension that is blocking the resource loading?

It could also be related to this: 
https://stackoverflow.com/a/42745512

**Comment by SimplyOnMyWay** [comment]
Thanks again @ofTheo - just tried that, with firefox. 
No joy again I'm afraid.

Haven't looked at the CORS angle just yet.

**Comment by ofTheo** [comment]
@SimplyOnMyWay for what it is worth your server hosted example works for me on Firefox and Chrome on macOS: 

![image](https://user-images.githubusercontent.com/144000/106523958-c51fce00-6496-11eb-9f10-8bdd014d5452.png)

I'll see if I can boot up my Ubunut machine and see if it runs there too.

**Comment by ofTheo** [comment]
Okay, I just tried with Ubuntu 20.04 and I also couldn't get it working in Firefox, but it worked with without any changes in Chrome. 

Can you try installing chrome and see if: `emrun --browser chrome bin/3DPrimitivesExample.html` works for you? 
I know that it is not an ideal solution but there might be something in the way firefox is handling security of downloads that breaks emscripten? 

**TLDR: on macOS both firefox and chrome work, but on Ubuntu just Chrome**

**Comment by SimplyOnMyWay** [comment]
Super @ofTheo that is working now! 
I'm also on Ubuntu 20.04.

I'll bear that in mind regarding firefox handling security of downloads.

Also, chrome throws a several warnings, which may shed some light.
The console states AudioContext was not allowed to start - this is the webaudio api as I understadn, which by the standard does not allow browsers to play audio without a user gesture. 
Also over 256 "WebGL texImage2D: no texture bound to target" warnings

Also (though not throwing warnings) the camera on my laptop was requested to be used by chrome.

The audio and camera operations seem strange, given it seems to application doesn't seem to require them - though I may well be missing this.

Fullscreen works when toggling via keyboard, but a console.log exception is thrown when mouse click on Fullscreen button.

Perhaps firefox is less tolerant to these warnings and interface requests with camera, audio? 


Not ideal as you say re Firefox on Ubuntu, but great to isolate the issue - many thanks!
I'll let folks know to open any work I share online in Chrome.

**Comment by ofTheo** [comment]
Hey. Glad to hear! 

That example actually does use the camera ( in one of the options to texture the shapes with the camera feed ). 
I had the same issue with fullscreen too. 

If you do track down any solutions to the above, please open an issue or PR and we'd be happy to try and get them patched in. 

Thanks for the update!
Theo
